db_sql.go

```diff
@@ -648,28 +649,97 @@ func (x *sqlSessionDB) Read(sessionkey string) (*Token, error) {
 			return nil, err
 		}
 	} else {
-		if err := token.Permit.Deserialize(epermit); err != nil {
-			return nil, err
+		t, err2, done := x.populateToken(token, epermit, userId, oauthID)
+		if done {
+			return t, err2
 		} else {
-			token.UserId = UserId(userId)
-			user, err := getUserFromUserId(x.db, token.UserId)
-			if err != nil {
-				return nil, err
-			}
-			token.Username = user.Username
-			token.Email = user.Email
-			if user.Type == UserTypeLDAP {
-				token.Identity = user.Username
+			return nil, nil
+		}
+	}
+}
+
+func (x *sqlSessionDB) populateToken(token *Token, epermit string, userId int64, oauthID sql.NullString) (*Token, error, bool) {
+	if err := token.Permit.Deserialize(epermit); err != nil {
+		return nil, err, true
+	} else {
+		token.UserId = UserId(userId)
+		user, err := getUserFromUserId(x.db, token.UserId)
+		if err != nil {
+			return nil, err, true
+		}
+		token.Username = user.Username
+		token.Email = user.Email
+		if user.Type == UserTypeLDAP {
+			token.Identity = user.Username
+		} else {
+			token.Identity = user.Email
+		}
+		if oauthID.Valid {
+			token.OAuthSessionID = oauthID.String
+		}
+
+		return token, nil, true
+	}
+	return nil, nil, false
```

Comment:
If I am understanding this code correctly, then I would remove the `else` and just have it's content be the rest of the function.